pkg/scheduler/framework/plugins/nodeports/node_ports.go

```diff
@@ -112,11 +114,59 @@ func getPreFilterState(cycleState *framework.CycleState) (preFilterState, error)
 func (pl *NodePorts) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
 		// Due to immutable fields `spec.containers[*].ports`, pod update events are ignored.
-		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
+		// TODO: Ideally, it's supposed to register only NodeCreated, because NodeUpdated event never means to have any free ports for the Pod.
+		// But, we may miss NodeCreated event due to preCheck.
+		// See: https://github.com/kubernetes/kubernetes/issues/109437
+		// And, we can remove NodeUpdated event once https://github.com/kubernetes/kubernetes/issues/110175 is solved.
```

Comment:
Is there already an ongoing PR for scheduling hints for node affinity?

With that implementation, maybe we can skip nodeaffinity precheck if the SchedulingHint feature gate is enabled. And we can remove `Update` here.

Reply:
> Is there already an ongoing PR for scheduling hints for node affinity?
@alculquicondor Yes, please review https://github.com/kubernetes/kubernetes/pull/119155 

> With that implementation, maybe we can skip nodeaffinity precheck if the SchedulingHint feature gate is enabled. And we can remove Update here.

@sanposhiho what do you think ?

Reply:
> maybe we can skip nodeaffinity precheck if the SchedulingHint feature gate is enabled. And we can remove Update here.

Yes, we can eliminate [NodeAffinity check](https://github.com/kubernetes/kubernetes/blob/10827a193a306ee33aebf84e110d2a2f0dee5d1c/pkg/scheduler/eventhandlers.go#L604) in preCheck, but the existing `preCheck` checks more than just NodeAffinity. 
https://github.com/kubernetes/kubernetes/blob/10827a193a306ee33aebf84e110d2a2f0dee5d1c/pkg/scheduler/eventhandlers.go#L572

For example, unready Node is created → the Pod failed by `nodeport` doesn't get informed of this `NodeCreated` event because preCheck filter out such events for Node which is tainted. → Node is now ready and untainted and this event should be delivered to plugins (unless other things in preCheck still reject it), **but as `NodeUpdate`**. This kind of scenarios could still happen. So, I'd suggest we keep it until we completely eliminate a whole preCheck.

Reply:
Right, so once we have the implementation for every plugin that is related to the existing prechecks, we can disable them based on the feature gate.

WDYT?

Regardless, we can proceed with this PR
/approve

Reply:
/hold cancel

Reply:
> Right, so once we have the implementation for every plugin that is related to the existing prechecks, we can disable them based on the feature gate.

Yes.